flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java

```diff
@@ -332,6 +334,22 @@ public String toString() {
 		return taskInfo.build();
 	}
 
+	/**
+	 * Get port keys representing the TM's configured endpoints. This includes mandatory TM endpoints such as
+	 * data and rpc as well as optionally configured endpoints for services such as prometheus reporter
+	 *
+	 * @return A deterministicly ordered Set of port keys to expose from the TM container
+	 */
+	private Set<String> getPortKeys() {
+		LinkedHashSet<String> tmPortKeys = new LinkedHashSet<>(Arrays.asList(TM_PORT_KEYS));
```

Comment:
Is this allowed to contain port ranges?

Reply:
Yes, ranges are allowed as value. In fact *anything* is allowed as value, because the implementation does not look at which value is assigned to a port key. It always overrules your assigned port value and replaces it with a single port provided by Mesos.

My understanding is, that port ranges in the first place are only used for one purpose: To allow picking a single available port from within the range. For example when running multiple TMs on the same machine, where ports would otherwise clash. But this complexity is not needed when spawning TMs with Mesos, because Mesos will always choose and assign a port, which is guarenteed to be available. Therefore, it's safe to only request a single port for each configured portKey.

Reply:
ah,I think i misunderstood earlier what the code does.

So what we _are_ doing is scanning the taskmanager config for port config options, gather the keys, and later these are overridden by mesos with some port?

If so, is there any way to restrict ports to certain ranges? For the taskmanager port this isn't important as for all intents and purposes it is only used internally anyway, but for reporters this port is very much user-facing.
In case of the PrometheusReporter this would imply that prometheus is always started _after_ the TM, as we can't determine the port beforehand. Is that correct?

Reply:
> So what we are doing is scanning the taskmanager config for port config options, gather the keys, and later these are overridden by mesos with some port?

Yes.

> In case of the PrometheusReporter this would imply that prometheus is always started after 
> the TM, as we can't determine the port beforehand. Is that correct?

No. Prometheus supports dynamicly reloading it's scrape configuration. So the other part of the monitoring solution is our Prometheus script, which periodically querries mesos to maintain an up-to-date list of active TM prometheus ports to scrape from. We're working with Mesosphere to open source this Prometheus scripting in DC/OS, but before that makes sense, we need Flink to let Mesos assign the scrape port to the TM container - hence this PR, which I hope will be quickly merged.